feat: improving manipulation demo

jmatejcz opened the pull request:
## Purpose
Improve performance of manipulation demo

## Proposed Changes
- `examples/manipulation-demo-streamlit.py` now launches everything - simulation, robo stack and UI. No need for second terminal.
- Changes to embodiment prompt to improve performance
- Added new scene configs 
- Now all configs that are available in the demo contains only  - red cubes, carrots and apples
Yellow cubes, blue cubes, tomatoes and corn were removed because our vision model do not recognize colors so these were easy to confuse
- Fixed a bug in connector where unique id was created during import time, which resulted in all connector nodes having the same name
- Fixed description of arguments in MoveObjectFromToToolInput
- Updated command in docs and added example prompts
- Using Vacuum Gripper which performs better in our demo (switched branch in demos.repos)

> [!WARNING]
If you already have rai-manipulaiton-demo cloned, VCS wont pull changes automatically. You need to either pull manually or remove the repo and clone again.


## Notes and Doubts

I didn't test it extensively on rai-perception refactor. It was developed and tested mainly when rai-openset-vision existed, so the tools were slightly different. But they are very similar so i' m betting that it works fine.

in experimental setup reset arm was added straight to agent after every tool call, so the arm is not stuck in weird positions. In production we obviously can't leave it like that. Reset arm tool can be added, but it will introduce more complexity to demo. 

I tried making adding launch methods to common file here https://github.com/RobotecAI/rai/pull/747/changes/a4f849580c5421828b1db6bc1914960e40fc21ac, but it seems like python package import do not work in launch files so I removed it. If you know if this can be done, please let me know 

If now one streamlit file launches everything, do we need the rest of the example mani files?
 
## Testing
Run manipulation demo, it should perform better


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->

## Summary by CodeRabbit

## Release Notes

* **New Features**
  * Added Streamlit-based interactive interface for the manipulation demo with predefined layout selection and dynamic scene configuration.
  * Integrated camera artifact support in chat interactions for multimodal AI responses.
  * Expanded predefined scene configurations for manipulation tasks.

* **Refactor**
  * Consolidated agent creation logic into a shared module for improved maintainability.
  * Centralized scene configuration path management for easier scenario setup.

* **Documentation**
  * Updated manipulation demo guide with simplified workflows and expanded usage examples.

* **Tests**
  * Added validation tests for connector naming and scene clearing functionality.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

boczekbartek committed:
feat(`manipulation_demo`): add constraints and rules to embodiment

boczekbartek committed:
feat: make sure that configs are possible

boczekbartek committed:
feat(`rai_bench`): update layouts

boczekbartek committed:
patch simulation_bridge

jmatejcz committed:
refactor: adjust tool descriptions

jmatejcz committed:
feat: add new scene configs

jmatejcz committed:
feat: add new configs without corn

jmatejcz committed:
refactor: adjust the configs

jmatejcz committed:
refactor: adjust the prompts

jmatejcz committed:
fix: config typos

jmatejcz committed:
fix: not defining node name at import time

jmatejcz committed:
feat: manipulation demo streamlit launches everything
added common file in examples
>
> Co-authored-by: bboczek <bartłomiej.boczek@robotec.ai>

jmatejcz committed:
refactor: return camera tool from create agent

jmatejcz committed:
fix: move launch method back to launch files

jmatejcz committed:
style: format

jmatejcz committed:
refactor:  remove redundant rclpy init

jmatejcz committed:
chore: use vacum gripper branch

jmatejcz committed:
docs: update docs

jmatejcz committed:
style: add back comment

jmatejcz committed:
style: comments and license

jmatejcz committed:
chore: version bump

jmatejcz committed:
docs: add examples of prompts

jmatejcz committed:
docs: add simbench install to mani demo

jmatejcz committed:
style: remove emojis

jmatejcz committed:
tests: adjust sim tests to changes

Juliaj commented on src/rai_core/rai/tools/ros2/manipulation/custom.py:
```diff
@@ -128,9 +128,9 @@ def _run(
-    z: float = Field(description="The z coordinate of the point to move to")
+    x: float = Field(description="The x coordinate of the point to move from")
+    y: float = Field(description="The y coordinate of the point to move from")
+    z: float = Field(description="The z coordinate of the point to move from")
```
What do you think about these ?

```suggestion

    x: float = Field(description="The x coordinate of the source point (pick-up location) in the manipulator frame")
    y: float = Field(description="The y coordinate of the source point (pick-up location) in the manipulator frame")
    z: float = Field(description="The z coordinate of the source point (pick-up location) in the manipulator frame")
    x1: float = Field(description="The x coordinate of the destination point (drop-off location) in the manipulator frame")
    y1: float = Field(description="The y coordinate of the destination point (drop-off location) in the manipulator frame")
    z1: float = Field(description="The z coordinate of the destination point (drop-off location) in the manipulator frame")

```
The thought about adding "in the manipulator frame" to the coordinate descriptions is to clarify the required coordinate system and help the LLM agents chain tools correctly. So when GetObjectPositionsTool returns positions in the target/manipulator frame, the agent can use them directly with MoveObjectFromToTool without confusion.

jmatejcz replied:
i didn't encounter problems where llm confuses 1st point with 2nd point, but it looks like a good change

Juliaj commented on src/rai_sim/rai_sim/simulation_bridge.py:
```diff
@@ -127,24 +127,28 @@ def load_base_config(cls, base_config_path: Path) -> "SceneConfig":
             )
-            for entity in content["entities"]
-        ]
+            # Ensure the object has the correct _prefix field
```
```suggestion
            # Set the ROS2 message type namespace 
```

jmatejcz replied:
deleted 437f25d2a9247ce7bbcb0d5b18df2a59c966a2c5

Juliaj commented on src/rai_sim/rai_sim/simulation_bridge.py:
```diff
@@ -127,24 +127,28 @@ def load_base_config(cls, base_config_path: Path) -> "SceneConfig":
-            for entity in content["entities"]
-        ]
+            # Ensure the object has the correct _prefix field
+            pose_stamped._prefix = "geometry_msgs/msg"
```
In theory, this is not needed because the _prefix is already set in `BaseGeometryModel`. Were you seeing runtime failures ?

jmatejcz replied:
I asked @boczekbartek , but it was long ago and he do not remember why he added this, but we agreed that it is probably not needed as _prefix has default value ( as you mentioned), so i  deleted that line 437f25d2a9247ce7bbcb0d5b18df2a59c966a2c5

Juliaj commented on examples/embodiments/manipulation_embodiment.json:
```diff
@@ -1,14 +1,22 @@
     "rules": [
-        "Always follow the specified payload limit; do not exceed 3 kg."
+        "Always follow the specified payload limit; do not exceed 3 kg.",
+        "Before starting the task, by looking at the image, determine if the task can be completed, of yes - proceed without asking. Otherwise report.",
```
```suggestion
        "Before starting the task, by looking at the image, determine if the task can be completed. If yes - proceed without asking. Otherwise report.",
```

jmatejcz replied:
applied 62e5985b321485c0eeea795f1ef37cd09d4a4522

Juliaj commented on examples/embodiments/manipulation_embodiment.json:
```diff
@@ -1,14 +1,22 @@
-        "Always follow the specified payload limit; do not exceed 3 kg."
+        "Always follow the specified payload limit; do not exceed 3 kg.",
+        "Before starting the task, by looking at the image, determine if the task can be completed, of yes - proceed without asking. Otherwise report.",
+        "Cubes height is 5cm. Remember to adjust z coordinate while stacking.",
```
```suggestion
        "Cubes are 5cm tall. When stacking, add 5cm to the z coordinate for each additional cube.",
```

jmatejcz replied:
applied 62e5985b321485c0eeea795f1ef37cd09d4a4522

Juliaj commented on examples/embodiments/manipulation_embodiment.json:
```diff
@@ -1,14 +1,22 @@
+        "Before starting the task, by looking at the image, determine if the task can be completed, of yes - proceed without asking. Otherwise report.",
+        "Cubes height is 5cm. Remember to adjust z coordinate while stacking.",
+        "Remember to not put objects on top of other objects, unless specifically instructed otherwise",
+        "Try to consider the typical size og objects when placing them (width, height, length)",
```
```suggestion
        "Try to consider the typical size of objects when placing them (width, height, length)",
```

jmatejcz replied:
applied 62e5985b321485c0eeea795f1ef37cd09d4a4522

Juliaj commented on examples/embodiments/manipulation_embodiment.json:
```diff
@@ -1,14 +1,22 @@
+        "Remember to not put objects on top of other objects, unless specifically instructed otherwise",
+        "Try to consider the typical size og objects when placing them (width, height, length)",
+        "Always leave at least a 5cm margin (x and y) between objects unless specifically instructed otherwise",
+        "If you repeat the task, make sure to run detection again, to check objects positions",
```
```suggestion
        "If you repeat the task, make sure to run detection again to check object positions",
```

jmatejcz replied:
applied 62e5985b321485c0eeea795f1ef37cd09d4a4522

Juliaj commented on examples/embodiments/manipulation_embodiment.json:
```diff
@@ -1,14 +1,22 @@
         "850 mm reach suitable for a variety of industrial applications.",
         "Interfaces to detect and manipulate objects.",
-        "Coordinate system: x - front to back (positive is forward), y - left to right (positive is right), z - up to down (positive is up)."
+        "Coordinate system and reachable space: x - front to back (positive is forward, [0.3, 0.5]), y - left to right (positive is right, [-0.5, 0.5]), z - up to down (positive is up)."
```
```suggestion
        "Coordinate system: x-axis (front to back, positive is forward, reachable range [0.3, 0.5] m), y-axis (left to right, positive is right, reachable range [-0.5, 0.5] m), z-axis (vertical, positive is up)."
 
```

jmatejcz replied:
applied 62e5985b321485c0eeea795f1ef37cd09d4a4522

Juliaj commented:
> I tried making adding launch methods to common file here https://github.com/RobotecAI/rai/commit/a4f849580c5421828b1db6bc1914960e40fc21ac, but it seems like python package import do not work in launch files so I removed it. 

What error were you seeing? My guess is that the launch file was run in a different Python context, so examples/ might not be on sys.path .

in the launch file, you could try following 

```
import sys
from pathlib import Path

# Add examples directory to path
examples_path = Path(__file__).parent.parent / "examples"
sys.path.insert(0, str(examples_path))

from manipulation_common import launch_description  # or whatever function name
```

jmatejcz committed:
refactor: remove redundant assign

jmatejcz committed:
refactor: small adjustiment to embodiment prompt

jmatejcz committed:
refactor: adjust the arg to tool descriptions

jmatejcz commented:
> > I tried making adding launch methods to common file here [a4f8495](https://github.com/RobotecAI/rai/commit/a4f849580c5421828b1db6bc1914960e40fc21ac), but it seems like python package import do not work in launch files so I removed it.
> 
> What error were you seeing? My guess is that the launch file was run in a different Python context, so examples/ might not be on sys.path .
> 
> in the launch file, you could try following
> 
> ```
> import sys
> from pathlib import Path
> 
> # Add examples directory to path
> examples_path = Path(__file__).parent.parent / "examples"
> sys.path.insert(0, str(examples_path))
> 
> from manipulation_common import launch_description  # or whatever function name
> ```

I get import error. I tried these approaches with adding path, but it seems like when you use `ros2 launch` python modules are not recognized properly

jmatejcz commented:
## PERFORMANCE TESTS 


I tried simple prompts that require 1 tool call and some more advanced , for example these (which are in docs now):
- "Place every apple on top of the cube"
-  "Stack all cubes into a tower"
-  "Arrange objects in a line"
- "Move cubes to the left side of the table"

Problems that we encountered on old version:
- Objects were falling from the gripper mid manipulation
- Positions of cubes were detected wrongly , tomatoes with apples and corn was just hard to identify 
- Arm was bumping into other objects 
- Arm was not behaving properly on the ends of table
- Agent struggled with understanding objects sizes (bad alignments and stacking, not leaving any margin) 

on the above prompts demo had ~50% acc.

We found out that vision model do not recognize colors -> so I left in demo only configs with red cubes, removed tomatoes and corn.

We decided that this should be showcase of agentic Ai, not a robot capability, so we replaced real gripper with vacuum one ( which is not adequate to what you see in simulation). We removed collisions of gripper's fingers. @knicked  did all that in https://github.com/RobotecAI/rai-manipulation-demo.

We changed the embodiment prompt so that agent understand the sizes of objects, leaves margins, etc. We also added  the arm range in prompt and made configs so that the objects do not exceed the range.

The results:
- Easy prompts that require 1 tool call -> almost 100% acc
- Harder ones (these above) -> 80% , after follow app prompts from user even 95%

Juliaj commented on examples/manipulation-demo-streamlit.py:
```diff
@@ -12,34 +12,272 @@
+    "2rc_2a",
+    "3rc_2a_1carrot",
+    "3carrots_3a_2rc",
+]
```
Please consider making this as a dict with values as layout descriptions (i.e. three red cubes) . The keys are too cryptic for the user to select as a layout.

jmatejcz replied:
good point , added c2a8cc3d07dd5308cdaf90aa0d3a33e2241cbe34

Juliaj commented on src/rai_core/rai/communication/ros2/connectors/base.py:
```diff
@@ -122,7 +122,8 @@ def __init__(
         super().__init__()
-
+        if node_name is None:
+            node_name = f"rai_ros2_connector_{str(uuid.uuid4())[-12:]}"
```
Adding a unit test for this is worthwhile. The unit tests enabled a safe code refactoring.

jmatejcz replied:
added 414c328602f7d357eac6a8229d235555d619b0e4

jmatejcz committed:
style: full scenario names in UI

jmatejcz committed:
tests: test different names of 2 connectors

jmatejcz commented:
@coderabbitai review

jmatejcz committed:
style: typos

Cytrus14 commented:
@jmatejcz 
I've noticed that the manipulation demo will not work without installing the `simbench` dependency group (`poetry install --with simbench`)  due missing `inflect` module:

```
Traceback (most recent call last):
  File "/home/user/rai/examples/manipulation-demo-streamlit.py", line 31, in <module>
    from rai_bench.manipulation_o3de import get_scenarios
  File "/home/user/rai/src/rai_bench/rai_bench/__init__.py", line 14, in <module>
    from .test_models import (
  File "/home/user/rai/src/rai_bench/rai_bench/test_models.py", line 23, in <module>
    import rai_bench.tool_calling_agent as tool_calling_agent
  File "/home/user/rai/src/rai_bench/rai_bench/tool_calling_agent/__init__.py", line 16, in <module>
    from .predefined.tasks import get_tasks
  File "/home/user/rai/src/rai_bench/rai_bench/tool_calling_agent/predefined/__init__.py", line 17, in <module>
    from .manipulation_tasks import get_manipulation_tasks
  File "/home/user/rai/src/rai_bench/rai_bench/tool_calling_agent/predefined/manipulation_tasks.py", line 24, in <module>
    from rai_bench.tool_calling_agent.tasks.manipulation import (
  File "/home/user/rai/src/rai_bench/rai_bench/tool_calling_agent/tasks/manipulation.py", line 18, in <module>
    import inflect
ModuleNotFoundError: No module named 'inflect'
```

jmatejcz committed:
fix: pass executor type properly

jmatejcz committed:
fix:: remove unreachable code

jmatejcz committed:
fix: use proper apple prefab in one of the configs

jmatejcz committed:
fix: remove redundant config

jmatejcz committed:
docs: adjust example prompt

jmatejcz commented:
> @jmatejcz I've noticed that the manipulation demo will not work without installing the `simbench` dependency group (`poetry install --with simbench`) due missing `inflect` module:
> 
> ```
> Traceback (most recent call last):
>   File "/home/user/rai/examples/manipulation-demo-streamlit.py", line 31, in <module>
>     from rai_bench.manipulation_o3de import get_scenarios
>   File "/home/user/rai/src/rai_bench/rai_bench/__init__.py", line 14, in <module>
>     from .test_models import (
>   File "/home/user/rai/src/rai_bench/rai_bench/test_models.py", line 23, in <module>
>     import rai_bench.tool_calling_agent as tool_calling_agent
>   File "/home/user/rai/src/rai_bench/rai_bench/tool_calling_agent/__init__.py", line 16, in <module>
>     from .predefined.tasks import get_tasks
>   File "/home/user/rai/src/rai_bench/rai_bench/tool_calling_agent/predefined/__init__.py", line 17, in <module>
>     from .manipulation_tasks import get_manipulation_tasks
>   File "/home/user/rai/src/rai_bench/rai_bench/tool_calling_agent/predefined/manipulation_tasks.py", line 24, in <module>
>     from rai_bench.tool_calling_agent.tasks.manipulation import (
>   File "/home/user/rai/src/rai_bench/rai_bench/tool_calling_agent/tasks/manipulation.py", line 18, in <module>
>     import inflect
> ModuleNotFoundError: No module named 'inflect'
> ```

I know, i noticed it and the commit is already present on the next PR #748

Cytrus14 commented:
Based on my tests with gpt-4o the manipulator demo does indeed perform better. However, there are still some shortcomings which we should keep in mind:
* The agent does not see objects obscured by the manipulator arm. This is expected, but may nonetheless lead to issues.
* Sometimes the arm will knock down other objects when moving.
* If the position of an object changes during a task (e.g. it is knocked-down by the arm), the agent will use the outdated object location.
 
Additionally, I believe that we should add a tool for resting the position of the arm to the agent. This will allow the agent to move the arm and allow it to see the obscured objects.

jmatejcz committed:
feat: add reseting arm

jmatejcz commented:
> Based on my tests with gpt-4o the manipulator demo does indeed perform better. However, there are still some shortcomings which we should keep in mind:
> 
>     * The agent does not see objects obscured by the manipulator arm. This is expected, but may nonetheless lead to issues.
> 
>     * Sometimes the arm will knock down other objects when moving.
> 
>     * If the position of an object changes during a task (e.g. it is knocked-down by the arm), the agent will use the outdated object location.
> 
> 
> Additionally, I believe that we should add a tool for resting the position of the arm to the agent. This will allow the agent to move the arm and allow it to see the obscured objects.

I added resetting arm tool
and added automatic reset after each move

Juliaj reviewed (approved):
LGTM.
